mithril-signer/src/mktree_store_sqlite.rs

```diff
@@ -0,0 +1,169 @@
+                primary key (position)
+            )",
+        )?;
+
+        Ok(connection)
+    }
+
+    fn get_element_at_position(&self, position: u64) -> StdResult<Option<Arc<MKTreeNode>>> {
+        let inner_store = self.inner_store.read().unwrap();
+        let query = "SELECT element FROM merkle_tree WHERE position = ?";
+        let mut statement = (*inner_store).prepare(query)?;
+        statement.bind((1, position as i64)).unwrap();
+        let result = if let Ok(sqlite::State::Row) = statement.next() {
+            Some(Arc::new(MKTreeNode::new(
+                statement.read::<Bytes, _>("element")?,
+            )))
+        } else {
+            None
+        };
+
+        Ok(result)
+    }
+
+    fn insert_elements_from_position(
+        &self,
+        position: u64,
+        elements: Vec<Arc<MKTreeNode>>,
+    ) -> StdResult<()> {
+        let inner_store = self.inner_store.read().unwrap();
+        let values_columns: Vec<&str> = repeat("(?, ?)").take(elements.len()).collect();
+        let values: Vec<sqlite::Value> =
+            elements
+                .into_iter()
+                .enumerate()
+                .fold(vec![], |mut vec, (i, elem)| {
+                    vec.append(&mut vec![
+                        sqlite::Value::Integer((position + i as u64) as i64),
+                        sqlite::Value::Binary((**elem).to_vec()),
+                    ]);
+                    vec
+                });
+        let query = format!(
+            "INSERT INTO merkle_tree(position, element) VALUES {}",
+            values_columns.join(", ")
+        );
+        let mut statement = (*inner_store).prepare(query)?;
+        statement.bind::<&[(_, sqlite::Value)]>(
+            values
+                .into_iter()
+                .enumerate()
+                .map(|(i, v)| (i + 1, v))
+                .collect::<Vec<_>>()
+                .as_slice(),
+        )?;
+        statement.next()?;
+
+        Ok(())
+    }
+}
+
+impl MMRStoreReadOps<Arc<MKTreeNode>> for MKTreeStoreSqlite {
+    fn get_elem(&self, pos: u64) -> MMRResult<Option<Arc<MKTreeNode>>> {
+        self.get_element_at_position(pos)
+            .with_context(|| {
+                format!("MKTreeStoreSqlite failed to retrieve element at position {pos}")
+            })
+            .map_err(|e| MMRError::StoreError(e.to_string()))
+    }
+}
+
+impl MMRStoreWriteOps<Arc<MKTreeNode>> for MKTreeStoreSqlite {
+    fn append(&mut self, pos: u64, elems: Vec<Arc<MKTreeNode>>) -> MMRResult<()> {
+        self.insert_elements_from_position(pos, elems)
+            .with_context(|| {
+                format!("MKTreeStoreSqlite failed to insert elements from position {pos}")
+            })
+            .map_err(|e| MMRError::StoreError(e.to_string()))
+    }
+}
+
+impl Clone for MKTreeStoreSqlite {
+    fn clone(&self) -> Self {
+        unimplemented!("Clone is not implemented for MKTreeStoreSqlite")
+    }
+}
+
+impl MKTreeLeafIndexer for MKTreeStoreSqlite {
```

Comment:
The MkTreeStoreSqlite is not a MkTreeLeafIndexer.
It would be better to have the trait only for MKTreeStoreInMemory and not allow to call `compute_proof` on a MKTree<MkTreeStoreSqlite>
This is probably a too complicated refactoring which also risks complicating the syntax.
This remains a design problem that we must remain vigilant about.

Reply:
You're right, I think that's a limitation of the design. 
However, the `MKTreeLeafIndexer` is only used when creating proofs which is not happening in the signer node.
In any case, I'll try to find a way to circumvent this problem in a different PR :+1: